generate_report_ui.gs

```diff
@@ -0,0 +1,263 @@
+/**
+ * Callback for rendering the view.
+ * @return {CardService.Card} The card to show to the user.
+ */
+function onHomePage(e) {
+  return createUI(e);
+}
+
+/**
+* Creates a UI and return the card
+* @return {CardService.Card} The card to show to the user
+*/
+function createUI(e, reportUrlVal) {
+  var i;
+  var card = CardService.newCardBuilder();
+  
+  // Add the tabs
+  var tabSection = CardService.newCardSection();
+  var generateReportAction = CardService.newAction().setFunctionName('createUI');
+  var btn1 = CardService.newTextButton().setText('Insights').setOnClickAction(generateReportAction);
+  var createRuleAction = CardService.newAction().setFunctionName('listRuleUI');
+  var btn2 = CardService.newTextButton().setText('Actions').setOnClickAction(createRuleAction);
+  var buttonSet = CardService.newButtonSet()
+    .addButton(btn1)
+    .addButton(btn2);
+  tabSection.addWidget(buttonSet);
+  
+  var getReportAction = CardService.newAction()
+      .setFunctionName('generateReport'); 
+  var getReport = CardService.newImageButton()
+    .setAltText("Generate Report")
+    .setIconUrl("https://www.gstatic.com/images/icons/material/system/1x/check_circle_black_48dp.png")
+    .setOnClickAction(getReportAction);
+  
+  var imageKeyValue = CardService.newKeyValue()
+  .setContent("<b>Generate Report</b>")
+  .setButton(getReport);
+  
+  tabSection.addWidget(imageKeyValue);
+  card.addSection(tabSection);
+ 
+  // Add the time Filter
+  var timeFilterSection = CardService.newCardSection().setCollapsible(true).setHeader("<b>TIME FILTER</b>"); 
+  var timeFilter = CardService.newSelectionInput()
+    .setType(CardService.SelectionInputType.RADIO_BUTTON)
+    .setFieldName("timeFilter");
+  
+  // add the items according to input
+  if(e.formInput) {
+    timeFilter.addItem("Last Hour", "LAST_HOUR", e.formInput.timeFilter == "LAST_HOUR");
+    timeFilter.addItem("Last 6 Hours", "LAST_6_HOUR", e.formInput.timeFilter == "LAST_6_HOUR");
+    timeFilter.addItem("Last 24 Hours", "LAST_24_HOUR", e.formInput.timeFilter == "LAST_24_HOUR");
+    timeFilter.addItem("Last 7 Days", "LAST_7_DAYS", e.formInput.timeFilter == "LAST_7_DAYS");
+    timeFilter.addItem("Last 30 Days", "LAST_30_DAYS", e.formInput.timeFilter == "LAST_30_DAYS");
+  }
+  else {
+    timeFilter.addItem("Last Hour", "LAST_HOUR", true);
+    timeFilter.addItem("Last 6 Hours", "LAST_6_HOUR", false);
+    timeFilter.addItem("Last 24 Hours", "LAST_24_HOUR", false);
+    timeFilter.addItem("Last 7 Days", "LAST_7_DAYS", false);
+    timeFilter.addItem("Last 30 Days", "LAST_30_DAYS", false);
+  }
```

Comment:
This code is duplicated. Can you create a function in which may be you can pass a map of filters you want to apply, selectionInput and default value.

Reply:
Card Service has certain limitations.
I tried doing it in function but wasn't successful

Reply:
Updated